### DISCUSSION TITLE
[v5] Best practice for resolving / importing objects registered in provider

### DISCUSSION THREAD

**Discussion by skrenek** [adonisjs/core#2647] [Help] [answered]
If I register a singleton in my app provider as such:

    this.app.container.singleton('MyApp/SomeSingleton', () => {
      return new SomeSingletonClass()
    })

Is this the best (only?) way to resolve that singleton in the app (for example, in a controller function)?

    const SomeSingleton = Application.container.resolveBinding('MyApp/SomeSingleton')

**Comment by RomainLanz** [comment, marked as answer]
Hey @skrenek! 👋 

It would be best to define its type like we are doing [here](https://github.com/adonisjs/lucid/blob/develop/adonis-typings/model.ts#L10).
Then, you will be able to import it via the IoC.

```ts
import SomeSingleton from '@ioc:MyApp/SomeSingleton'
```

--

Now, the main question is, do you need to create a provider?
The provider's goal is to remove the glue-code for an external package, not for internal purposes.

If you need a singleton, return an instance of a class.

```ts
class SomeSingleton {}

export default new SomeSingleton()
```

**Reply by skrenek** [reply]
Thanks @RomainLanz.  Your question is indeed a valid one, and normally I do as you suggested.  This is actually a smaller piece of a bigger problem I am encountering in creating a reusable Adonis package.  However with multiple unknowns I opted to ask this question on its own as a smaller piece of the puzzle.

Am I understanding this correctly then as stated below?

The IoC binding via `.bind` or `.singleton` provides the implementation of this statement at runtime:
`import SomeSingleton from '@ioc:MyApp/SomeSingleton'`

The `declare module '@ioc:MyApp/SomeSingleton'` declaration then (contracts internally, adonis-typings in a package) provides typescript with the info it needs to be happy during development.

**Reply by thetutlage** [reply]
Yup. So, `Ioc.bind` adds a binding to the container, which TypeScript doesn't know about. So we need to tell TypeScript separately about it by using [declare module](https://www.typescriptlang.org/docs/handbook/modules.html#ambient-modules)
